packages/next/src/build/index.ts

```diff
@@ -1239,6 +1240,17 @@ export default async function build(
 
       process.env.NEXT_PHASE = PHASE_PRODUCTION_BUILD
 
+      // We limit the number of workers used based on the number of CPUs and
+      // the current available memory. This is to prevent the system from
+      // running out of memory as well as maximize speed. We assume that
+      // each worker will consume ~1GB of memory in a production build.
+      // For example, if the system has 10 CPU cores and 8GB of remaining memory
+      // we will use 8 workers.
+      const numWorkers = Math.min(
+        config.experimental.cpus || 1,
+        Math.floor(os.freemem() / 1e9)
+      )
```

Comment:
Maybe we should still allow `config.experimental.cpus` to override even if it's more than the `os.freemem()` calc for testing

```suggestion
      const numWorkers =
        config.experimental.cpus !== defaultConfig.experimental.cpus
          ? config.experimental.cpus
          : Math.min(config.experimental.cpus || 1, Math.floor(os.freemem() / 1e9));

```